src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -812,101 +731,204 @@ public int getCurrentRunningCount() {
-     * @throws InterruptedException Thrown if this thread is interrupted while waiting to acquire a worker
-     */
-    public Worker getWorker() throws InterruptedException {
-      synchronized (workersLock) {
-        while (availableWorkers.isEmpty() && ! shutdownFinishing) {
-          if (currentPoolSize < maxPoolSize) {
-            return makeNewWorker();
-          } else {
-            waitingForWorker = true;
-            try {
-              workersLock.wait();
-            } finally {
-              waitingForWorker = false;
-            }
-          }
+      int casPoolSize;
+      while ((casPoolSize = currentPoolSize.get()) < maxPoolSize) {
+        if (currentPoolSize.compareAndSet(casPoolSize, casPoolSize + 1)) {
+          makeNewWorker();
         }
-        
-        return availableWorkers.poll();
       }
     }
     
     /**
-     * This function REQUIRES that workersLock is synchronized before calling.  This call creates 
-     * a new worker, starts it, but does NOT add it as an available worker (so you can immediately 
-     * use it).  If you want this worker to be available for other tasks, it must be added to the 
-     * {@code availableWorkers} queue.
-     * 
-     * @return Newly created worker, started and ready to accept work
+     * This call creates and starts a new worker.  It does not modify {@link currentPoolSize} so 
+     * that MUST be updated in a thread safe way before this is invoked.  As soon as the worker 
+     * starts it will attempt to start taking tasks, no further action is needed.
      */
-    protected Worker makeNewWorker() {
+    protected void makeNewWorker() {
       Worker w = new Worker(this, threadFactory);
-      currentPoolSize++;
       w.start();
-      
-      // will be added to available workers when done with first task
-      return w;
     }
     
     /**
-     * Shuts down the worker and ensures this now dead worker wont be used.
+     * Adds a worker to the head of the idle worker chain.
      * 
-     * @param w worker to shutdown
+     * @param worker Worker that is ready to become idle
      */
-    protected void killWorker(Worker w) {
-      // IMPORTANT** if this lock is removed, it is important to read the comment bellow
-      synchronized (workersLock) {
-        /* this will throw an exception if the worker has already stopped, 
-         * we want to ensure the pool size is not decremented more than once for a given worker.
-         * 
-         * We are able to stop first, because we are holding the workers lock.  In the future if we 
-         * try to reduce locking around here, we need to ensure that the worker is removed from the 
-         * available workers BEFORE stopping.
-         */
-        w.stopIfRunning();
-        currentPoolSize--;
-        // it may not always be here, but it sometimes can (for example when a worker is interrupted)
-        availableWorkers.remove(w);
+    protected void addWorkerToIdleChain(Worker worker) {
```

Comment:
We basically have idle workers as a linked list.  We have an atomic reference that we use for the initial thread safe operations, and all new idle workers are put on the front of the chain for that reason.

The only way a worker's reference should change is if it is is changing on the worker thread (ie the worker thread should be invoking into this function)...and the workers reference is only removed after it is removing itself from queue.